### PR TITLE
Support union types and optional fields with dot separation on UpdateData

### PR DESCRIPTION
**Pull request by thebrianchen**
Fixes a bug where `UpdateData` would not recognize optional, dot-separated string fields. See [comment](https://github.com/firebase/firebase-js-sdk/pull/5268#issuecomment-905948377) for repro details.

### CODE REVIEW

File: .changeset/clean-cameras-check.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@firebase/firestore': minor
+---
+
+Fixed a bug where UpdateData did not recognize optional, dot-separated string fields
```
Comment (egilmorez) [inline .changeset/clean-cameras-check.md:5]: If this goes straight into release notes, I'd suggest `UpdateData` as a literal, and a period.  Otherwise LGTM, thanks!

**Review by egilmorez — APPROVED** [review]

**Comment by svailla4** [comment]
How do we use updateData in firebase-admin? @thebrianchen

File: packages/firestore/src/lite-api/types.ts [hunk 2]
```diff
@@ -35,15 +35,26 @@ export type NestedUpdateFields<T extends Record<string, unknown>> =
   UnionToIntersection<
     {
       // Check that T[K] extends Record to only allow nesting for map values.
-      [K in keyof T & string]: T[K] extends Record<string, unknown>
-        ? // Recurse into the map and add the prefix in front of each key
-          // (e.g. Prefix 'bar.' to create: 'bar.baz' and 'bar.qux'.
-          AddPrefixToKeys<K, UpdateData<T[K]>>
-        : // TypedUpdateData is always a map of values.
-          never;
+      // Union with `undefined` to allow for optional nested fields
```
Comment (yuchenshi) [inline packages/firestore/src/lite-api/types.ts:38]: outdated

File: packages/firestore/test/lite/integration.test.ts [hunk 3]
```diff
@@ -1542,6 +1542,106 @@ describe('withConverter() support', () => {
...
+      });
+
+      it('supports union fields', () => {
+        interface TestObjectUnion {
+          optionalStr?: string;
+          nested?:
+            | {
+                requiredStr: string;
+              }
+            | { requiredStrObject: string };
```
Comment (yuchenshi) [inline packages/firestore/test/lite/integration.test.ts:1595]: May be a better example would be `: {foo: number}`?

Comment (thebrianchen) [reply packages/firestore/test/lite/integration.test.ts:1595]: Done.

File: packages/firestore/src/lite-api/types.ts (hunk 2 above)
Comment (thebrianchen) [reply packages/firestore/src/lite-api/types.ts:38]: Good catch, thanks

File: packages/firestore/test/lite/integration.test.ts [hunk 4]
```diff
@@ -1542,6 +1542,108 @@ describe('withConverter() support', () => {
         });
       });
 
+      it('supports optional fields', () => {
```
Comment (schmidt-sebastian) [inline packages/firestore/test/lite/integration.test.ts:1545]: Can you add a test for `null`?

File: packages/firestore/src/lite-api/types.ts [hunk 5]
```diff
@@ -34,16 +34,29 @@ export type Primitive = string | number | boolean | undefined | null;
...
+ * props) or `{a: A} | {b: B}`.
+ *
+ * In this use case, `T1` is used to distribute the union types of `T[K]` on
+ * `Record`, since `T[K]` is evaluated as an expression and not distributed.
+ *
+ * See https://www.typescriptlang.org/docs/handbook/advanced-types.html#distributive-conditional-types
+ */
+export type ChildUpdateFields<T1, K extends string> =
+  // Only allow nesting for map values
+  T1 extends Record<string, unknown>
```
Comment (schmidt-sebastian) [inline packages/firestore/src/lite-api/types.ts:53]: Is there a better name than T1? Reminds me of TI-89

**Review by schmidt-sebastian — APPROVED** [review]

File: packages/firestore/test/lite/integration.test.ts (hunk 4 above)
Comment (thebrianchen) [reply packages/firestore/test/lite/integration.test.ts:1545]: Done.

File: packages/firestore/src/lite-api/types.ts (hunk 5 above)
Comment (thebrianchen) [reply packages/firestore/src/lite-api/types.ts]: Renamed to `<K,V>`

File: packages/firestore/src/lite-api/reference.ts [hunk 6]
```diff
@@ -78,12 +78,9 @@ export type WithFieldValue<T> = T extends Primitive
  */
 export type UpdateData<T> = T extends Primitive
   ? T
-  : T extends Map<infer K, infer V>
```
Comment (Feiyang1) [inline packages/firestore/src/lite-api/reference.ts:81]: Why do you not need to handle `Map`s separately?

File: packages/firestore/src/lite-api/types.ts [hunk 7]
```diff
@@ -34,16 +34,29 @@ export type Primitive = string | number | boolean | undefined | null;
...
+ * Helper for calculating the nested fields for a given type T1. This is needed
+ * to distribute union types such as `undefined | {...}` (happens for optional
+ * props) or `{a: A} | {b: B}`.
+ *
+ * In this use case, `V` is used to distribute the union types of `T[K]` on
+ * `Record`, since `T[K]` is evaluated as an expression and not distributed.
+ *
+ * See https://www.typescriptlang.org/docs/handbook/advanced-types.html#distributive-conditional-types
+ */
+export type ChildUpdateFields<K extends string, V> =
```
Comment (Feiyang1) [inline packages/firestore/src/lite-api/types.ts:51]: This looks identical to the initial implementation. What has changed  that adds the support for optional fields?

File: packages/firestore/src/lite-api/reference.ts (hunk 6 above)
Comment (thebrianchen) [reply packages/firestore/src/lite-api/reference.ts:81]: `UpdateData` shouldn't allow `Maps` to begin with because it was designed to take `{ [fieldPath: string]: any}` as an input originally. `WithFieldValue` and `PartialWithFieldValue` both don't allow `Map`s.

File: packages/firestore/src/lite-api/types.ts (hunk 7 above)
Comment (thebrianchen) [reply packages/firestore/src/lite-api/types.ts:51]: This is an excerpt from the documentation link:
> Conditional types in which the checked type is a naked type parameter are called distributive conditional types. Distributive conditional types are automatically distributed over union types during instantiation. For example, an instantiation of T extends U ? X : Y with the type argument A | B | C for T is resolved as (A extends U ? X : Y) | (B extends U ? X : Y) | (C extends U ? X : Y).

Here's a TS playground [link](https://www.typescriptlang.org/play?#code/PTAEBUAsEsGdQC4E8AOBTUdQDsD22BaAEzgQCdoAjAVwTSNAENsGy0U3Y1sF4FIMaAB6lo2AOagABgDk0sOkQCqKIozoAxaGgA2RWFMSo0AOgBQydKA25cAHnChhdFvABKaAMa4yROwooJABpqbABrPAB3bAA+GNAAXlAAbzMAbQBpTGxQMLQkXAAzCFAAMlAAsXEAXQAuCEzqpyEXfVAPb19-ciqg0FCI3GiYs1Ax8bGAfgaM6tGJ8frsNAA3NDIzAF8AbjMzEAgYPmNMeGWFeiYWUBJKmjo+ARx5RVBC7T0+yJhPSFA2Djybi8RBPX7McQYAC2jCIGDEoIwAAU3BYTjZcAB1aD8TFkRgodBkBzNVruLw+PyVYLMJBxRIpeaZbK5fJFErlak1eoAEVIFHuaHAxgcjRGOzRVj5d1oQpFADV4kl5aTuG0OpTuoFxH0BlFYvMFqBpvLDQslqt1rt9mAVGo6Dz1IxQJEcX8xMsyKBcCgENB8IwdG8PkRzJYMHb1GhHQhGNj+AB5X3+7CBuQXIhaXR+JWMsbnRT1VITDg++pc0AAH36LDQ72WRHmmyrNbh9foW2tBygGAUzDUvkwUJQOjQUOB6hTThWgeoUfg4EaTHgzGagNgsBT9SkAApF7MW6E22J6ABKVWudoUroADSCAE0YoZA5FGEh4JxcDo1nxcNJlmsZBSDaFRiJ4GBSHkBTFDuR51ieDDVvu1SnoYWBSMkmxSOY3jYAoiCwAAomQZA+PUGJ2JGDpOvGkBJn6AY6OmihZnoubFs8GZFmapYoPUcHto2YybJ2ewHDuKpJMhh61oJ549jc-JULKDDQMOo7jjwk74NOs7zqAKqMCuikygx2B9Pi-DrKCq6rsI66bvg24qs4arkp0fh3o+wEHFgqx6a8RnbgJCEXuq16eQ+8TVjMTSuZeGq3lFUhfD87rwDOOjQKGZi4fheDEaRZDkbYtF4gSRKUaoUYxnGbr0SmaYvPQrFEOx8wFvQ3ElqRfGtvBDZNqJQA) that shows the before and after.

Comment (Feiyang1) [reply packages/firestore/src/lite-api/types.ts:51]: Thanks for the example! Now it makes sense to me.

**Review by Feiyang1 — APPROVED** [review]
